File: apps/client-macos/Sources/TimeTrack/Storage/ImageBufferStore.swift
import Foundation

/// The multi-display grouping stamped on a buffered capture: which tick it belongs to, and where
/// it sat in that tick. Nil for records written by a build older than multi-display capture.
struct CaptureGroup: Equatable {
    let id: String
    let displayIndex: Int
    let displayCount: Int
}

/// PRD §6.2 / §7.4 — durable, file-backed binary buffer for captured screenshots. Mirrors
/// `BufferStore`'s durability model (atomic write-tmp-then-rename, startup sweep of `.tmp-*`,
/// one file per record) but stores raw JPEG bytes, not JSON — screenshots are binary multipart
/// (2.2a §13), so they get their own store and their own uploader/engine. The filename
/// `<capturedAtMillis>__<uuidv7>__<groupId>__<displayIndex>__<displayCount>.jpg` carries FIFO
/// order, identity (the server `id`), the capture time (half the server PK `[id, timestamp]` +
/// the partition key), AND the multi-display grouping — all WITHOUT reading the file, so the
/// capture time is stamped once and reused verbatim on every upload retry.
///
/// The two-component form `<capturedAtMillis>__<uuidv7>.jpg` is still parsed: it is what the
/// build currently on pilot Macs writes, and those pending captures must still drain after an
/// update rather than being silently dropped on the floor. They carry no group, which is
/// accurate — they were single main-display grabs.
final class ImageBufferStore {
    static let shared = ImageBufferStore(directory: ImageBufferStore.defaultDirectory())

    private let directory: URL
    private let clock: () -> Date

    init(directory: URL, clock: @escaping () -> Date = Date.init) {
        self.directory = directory
        self.clock = clock
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        sweepTemporaries()
    }

    /// ~/Library/Application Support/<container>/screenshots/ — the container is per-install, so
    /// a dev build never drains the released app's pending captures. See `AppInstall`.
    static func defaultDirectory() -> URL {
        AppInstall.supportDirectory("screenshots")
    }

    /// Atomic enqueue: write `.tmp-<id>`, then rename to the encoded record name.
    ///
    /// `group` is nil only for a capture with no display context to record. Every current caller
    /// passes one; the parameter stays optional so a record can still be written (and drained) in
    /// the shape the previous build used.
    func enqueue(id: String, capturedAt: Date, jpeg: Data, group: CaptureGroup? = nil) {
        let millis = Int64(capturedAt.timeIntervalSince1970 * 1000)
        let dst = directory.appendingPathComponent(Self.filename(millis: millis, id: id, group: group))
        let tmp = directory.appendingPathComponent(".tmp-\(id)")
        do {
            try jpeg.write(to: tmp, options: .atomic)
            try? FileManager.default.removeItem(at: dst)
            try FileManager.default.moveItem(at: tmp, to: dst)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
        }
    }

    /// FIFO (capture time asc) records, up to `limit`. Returns file URLs — callers read bytes lazily.
    func take(limit: Int) -> [(id: String, capturedAt: Date, group: CaptureGroup?, url: URL)] {
        allRecords().prefix(limit).map {
            (id: $0.id,
             capturedAt: Date(timeIntervalSince1970: Double($0.capturedAtMillis) / 1000),
             group: $0.group,
             url: $0.url)
        }
    }

    /// Screenshots still waiting to upload. Directory listing only, no image is read.
    func pendingCount() -> Int {
        take(limit: Int.max).count
    }

    func remove(id: String) {
        for rec in allRecords() where rec.id == id {
            try? FileManager.default.removeItem(at: rec.url)
        }
    }

    /// Age-bound (drop older than `maxAge`) then count-bound (trim oldest beyond `maxCount`).
    /// Images are large, so unlike the JSON buffer this also caps by count.
    func prune(olderThan maxAge: TimeInterval, maxCount: Int) {
        let cutoff = Int64(clock().addingTimeInterval(-maxAge).timeIntervalSince1970 * 1000)
        var records = allRecords()
        for rec in records where rec.capturedAtMillis < cutoff {
            try? FileManager.default.removeItem(at: rec.url)
        }
        records = records.filter { $0.capturedAtMillis >= cutoff }
        if records.count > maxCount {
            for rec in records.prefix(records.count - maxCount) {
                try? FileManager.default.removeItem(at: rec.url)
            }
        }
    }

    func clear() {
        for rec in allRecords() { try? FileManager.default.removeItem(at: rec.url) }
    }

    // MARK: - internals

    private struct Record {
        let url: URL; let capturedAtMillis: Int64; let id: String; let group: CaptureGroup?
    }

    /// The record name. Underscore-pair separated; a UUID contains no `__`, so the components
    /// can never be ambiguous.
    static func filename(millis: Int64, id: String, group: CaptureGroup?) -> String {
        guard let group else { return "\(millis)__\(id).jpg" }
        return "\(millis)__\(id)__\(group.id)__\(group.displayIndex)__\(group.displayCount).jpg"
    }

    private func contents() -> [URL] {
        (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)) ?? []
    }

    private func parse(_ url: URL) -> Record? {
        let name = url.lastPathComponent
        guard name.hasSuffix(".jpg") else { return nil }
        let parts = String(name.dropLast(4)).components(separatedBy: "__")  // strip ".jpg"
        guard let millis = Int64(parts.first ?? "") else { return nil }
        // Two components: written by a build older than multi-display capture. Still drainable,
        // and genuinely ungrouped — it was a single main-display grab.
        if parts.count == 2 {
            return Record(url: url, capturedAtMillis: millis, id: parts[1], group: nil)
        }
        guard parts.count == 5, let index = Int(parts[3]), let count = Int(parts[4]) else {
            return nil
        }
        return Record(url: url, capturedAtMillis: millis, id: parts[1],
                      group: CaptureGroup(id: parts[2], displayIndex: index, displayCount: count))
    }

    private func allRecords() -> [Record] {
        contents().compactMap(parse)
            .sorted { ($0.capturedAtMillis, $0.id) < ($1.capturedAtMillis, $1.id) }
    }

    private func sweepTemporaries() {
        for url in contents() where url.lastPathComponent.hasPrefix(".tmp-") {
            try? FileManager.default.removeItem(at: url)
        }
    }
}

File: apps/client-windows/src/NiftyTimer/Storage/ImageBufferStore.cs
using System.Globalization;

namespace NiftyTimer.Storage;

/// <summary>
/// Which tick a buffered capture belongs to, and where it sat within that tick. Stamped on every
/// screenshot so the dashboard can show a multi-monitor desk as one moment rather than as
/// unrelated images.
///
/// <c>DisplayIndex</c> and <c>DisplayCount</c> are bounded by the server schema
/// (<c>displayIndex</c> 0–15, <c>displayCount</c> 1–16). Exceeding either is a 422, which
/// classifies as permanent — the capture would be dropped rather than retried — so the grabber
/// caps the number of displays it reports rather than letting an unusual desk silently lose its
/// screenshots.
/// </summary>
public sealed record CaptureGroup(string Id, int DisplayIndex, int DisplayCount)
{
    /// <summary>Server bound: <c>displayCount: z.coerce.number().int().min(1).max(16)</c>.</summary>
    public const int MaxDisplays = 16;
}

/// <summary>One buffered capture, identified without opening the file.</summary>
public sealed record BufferedImage(string Id, DateTimeOffset CapturedAt, CaptureGroup Group, string Path);

/// <summary>The image-buffer seam. Tests substitute a spy.</summary>
public interface IImageBuffer
{
    void Enqueue(string id, DateTimeOffset capturedAt, byte[] jpeg, CaptureGroup group);

    IReadOnlyList<BufferedImage> Take(int limit);

    void Remove(string id);

    void Prune(TimeSpan maxAge, int maxCount);

    void Clear();

    int PendingCount();
}

/// <summary>
/// PRD §6.2 / §7.4 — durable, file-backed buffer for captured screenshots.
///
/// Same durability model as <see cref="BufferStore"/> (one file per record, write-then-rename,
/// startup sweep of <c>.tmp-*</c>) but it holds raw JPEG bytes rather than JSON, because
/// screenshots upload as binary multipart through their own uploader and engine.
///
/// The filename
/// <c>&lt;capturedAtMillis&gt;__&lt;id&gt;__&lt;groupId&gt;__&lt;displayIndex&gt;__&lt;displayCount&gt;.jpg</c>
/// carries FIFO order, the server <c>id</c>, the capture time and the whole display grouping — so
/// none of it needs the file to be read, and in particular the capture time is stamped ONCE and
/// reused verbatim on every retry. That matters more than it looks: <c>timestamp</c> is half the
/// server's composite primary key <c>[id, timestamp]</c> and also its monthly partition key, so a
/// retry that recomputed "now" would land in a different partition under a different key and
/// duplicate the row instead of upserting it.
///
/// A UUID contains no <c>__</c>, so the five components can never be ambiguous.
///
/// Not a capture path — it stores what capture already produced — so it takes no gate and lives
/// in <c>NiftyTimer.Storage</c>.
/// </summary>
public sealed class ImageBufferStore : IImageBuffer
{
    private readonly string _directory;
    private readonly Func<DateTimeOffset> _clock;

    public ImageBufferStore(string directory, Func<DateTimeOffset>? clock = null)
    {
        _directory = directory;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        Directory.CreateDirectory(directory);
        SweepTemporaries();
    }

    public void Enqueue(string id, DateTimeOffset capturedAt, byte[] jpeg, CaptureGroup group)
    {
        var millis = capturedAt.ToUnixTimeMilliseconds();
        var dst = Path.Combine(_directory, FileName(millis, id, group));
        var tmp = Path.Combine(_directory, $".tmp-{id}");
        try
        {
            File.WriteAllBytes(tmp, jpeg);
            File.Move(tmp, dst, overwrite: true);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            TryDelete(tmp);
        }
    }

    /// <summary>
    /// FIFO (capture time ascending), up to <paramref name="limit"/>. Returns paths — the caller
    /// reads the bytes lazily, so listing the buffer for the menu's pending count never touches
    /// a single image.
    /// </summary>
    public IReadOnlyList<BufferedImage> Take(int limit)
    {
        var result = new List<BufferedImage>();
        foreach (var rec in AllRecords())
        {
            if (result.Count >= limit)
            {
                break;
            }

            result.Add(new BufferedImage(
                rec.Id,
                DateTimeOffset.FromUnixTimeMilliseconds(rec.CapturedAtMillis),
                rec.Group,
                rec.Path));
        }

        return result;
    }

    public void Remove(string id)
    {
        foreach (var rec in AllRecords())
        {
            if (rec.Id == id)
            {
                TryDelete(rec.Path);
            }
        }
    }

    /// <summary>
    /// Age-bound then count-bound. Unlike the JSON buffers this one is also capped by count,
    /// because a week of unsent screenshots from a laptop that has been offline is measured in
    /// gigabytes, not kilobytes.
    /// </summary>
    public void Prune(TimeSpan maxAge, int maxCount)
    {
        var cutoff = _clock().Subtract(maxAge).ToUnixTimeMilliseconds();
        var surviving = new List<Record>();
        foreach (var rec in AllRecords())
        {
            if (rec.CapturedAtMillis < cutoff)
            {
                TryDelete(rec.Path);
            }
            else
            {
                surviving.Add(rec);
            }
        }

        for (var i = 0; i < surviving.Count - maxCount; i++)
        {
            TryDelete(surviving[i].Path);
        }
    }

    public void Clear()
    {
        foreach (var rec in AllRecords())
        {
            TryDelete(rec.Path);
        }
    }

    /// <summary>Captures still waiting to upload. Directory listing only; no image is read.</summary>
    public int PendingCount() => AllRecords().Count;

    internal static string FileName(long millis, string id, CaptureGroup group) =>
        string.Create(
            CultureInfo.InvariantCulture,
            $"{millis}__{id}__{group.Id}__{group.DisplayIndex}__{group.DisplayCount}.jpg");

    private readonly record struct Record(string Path, long CapturedAtMillis, string Id, CaptureGroup Group);

    private string[] Contents()
    {
        try
        {
            return Directory.GetFiles(_directory);
        }
        catch (DirectoryNotFoundException)
        {
            return [];
        }
    }

    private static Record? Parse(string path)
    {
        var name = Path.GetFileName(path);
        if (!name.EndsWith(".jpg", StringComparison.Ordinal))
        {
            return null;
        }

        var parts = name[..^4].Split("__");
        if (parts.Length != 5 ||
            !long.TryParse(parts[0], NumberStyles.Integer, CultureInfo.InvariantCulture, out var millis) ||
            !int.TryParse(parts[3], NumberStyles.Integer, CultureInfo.InvariantCulture, out var index) ||
            !int.TryParse(parts[4], NumberStyles.Integer, CultureInfo.InvariantCulture, out var count))
        {
            return null;
        }

        return new Record(path, millis, parts[1], new CaptureGroup(parts[2], index, count));
    }

    private List<Record> AllRecords()
    {
        var records = new List<Record>();
        foreach (var path in Contents())
        {
            if (Parse(path) is { } rec)
            {
                records.Add(rec);
            }
        }

        records.Sort(static (a, b) =>
        {
            var byTime = a.CapturedAtMillis.CompareTo(b.CapturedAtMillis);
            return byTime != 0 ? byTime : string.CompareOrdinal(a.Id, b.Id);
        });
        return records;
    }

    private void SweepTemporaries()
    {
        foreach (var path in Contents())
        {
            if (Path.GetFileName(path).StartsWith(".tmp-", StringComparison.Ordinal))
            {
                TryDelete(path);
            }
        }
    }

    private static void TryDelete(string path)
    {
        try
        {
            File.Delete(path);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            // Another pass got there first, or the file is briefly locked; the next cycle retries.
        }
    }
}
